Merge ProfiledTensorType and TensorType

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24284 Merge ProfiledTensorType and TensorType**

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

Differential Revision: [D16794034](https://our.internmc.facebook.com/intern/diff/D16794034)

zdevito committed:
Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

zdevito committed:
Update on "Merge ProfiledTensorType and TensorType"

Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

gh-metadata: pytorch pytorch 24284 gh/zdevito/85/head

zdevito committed:
Update on "Merge ProfiledTensorType and TensorType"

Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

gh-metadata: pytorch pytorch 24284 gh/zdevito/85/head

zdevito committed:
Update on "Merge ProfiledTensorType and TensorType"

Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

gh-metadata: pytorch pytorch 24284 gh/zdevito/85/head

zdevito committed:
Update on "Merge ProfiledTensorType and TensorType"

Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

gh-metadata: pytorch pytorch 24284 gh/zdevito/85/head

zdevito committed:
Update on "Merge ProfiledTensorType and TensorType"

Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

gh-metadata: pytorch pytorch 24284 gh/zdevito/85/head

Krovatkin commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -294,58 +265,67 @@ inline c10::optional<T> merge_primitive(
+  using ListOfOptionalInts = std::vector<c10::optional<int64_t>>;
   VaryingShape(const std::vector<int64_t>& vec)
-      : size_(vec.size()), dims_(vec.begin(), vec.end()) {}
+      : VaryingShape(ListOfOptionalInts(vec.begin(), vec.end())) {}
```
Errr, deja vu, isn't there another PR floating with this exact change?

zdevito replied:
yeah, I split that out from this one, because it is higher priority to merge that one to prevent a hang in some client code. I will delete these changes from this PR.

Krovatkin reviewed (approved):
:shipit:

zdevito commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -1105,51 +1100,53 @@ class ShapePropagator {
+                "aten::prod(Tensor self, *, int? dtype) -> Tensor",
+            },
+            [](Node* node) -> type_vec_t {
+              if (auto type = node->input(0)->type()->cast<TensorType>()) {
```
yes, seems like that is true

zdevito committed:
Update on "Merge ProfiledTensorType and TensorType"

Merge ProfiledTensorType and TensorType

This PR finishes the unification of all Tensor types into a single object.
ProfiledTensorType is renamed to TensorType and the old TensorType is
deleted.

Notes:
* Fixes bug in merge for VaryingShape by changing its representation to an
 optional list of optional ints.
* Removes ProfiledTensorType::create(type) invocations that can now
  simply be expect calls on tensor type.

gh-metadata: pytorch pytorch 24284 gh/zdevito/85/head
